### PR TITLE
CPO-14 docker config

### PR DESCRIPTION
**Pull request by smathangi**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Review by tomxelliott — APPROVED** [review]

File: cpo-docker/compose/cpo.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      CPO_S2S_AUTHORISED_SERVICES: xui_webapp
+    ports:
+      - 4457:4457
+    depends_on:
+      - cpo-db
+    networks:
+      - ccd-network
+
+  cpo-db:
+    image: postgres:9.6
```
Comment (sullivangareth) [inline cpo-docker/compose/cpo.yml:23]: Should be version 11 ?

File: docker-compose.yml [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
+      IDAM_API_URL: https://idam-api.aat.platform.hmcts.net
+      S2S_URL: http://rpe-service-auth-provider-aat.service.core-compute-aat.internal
     ports:
-      - $SERVER_PORT:$SERVER_PORT
+      - 4457:4457
+    depends_on:
+      - cpo-db
+
+  cpo-db:
+    image: postgres:9.6
```
Comment (sullivangareth) [inline docker-compose.yml:20]: version 11

File: cpo-docker/compose/cpo.yml (hunk 1 above)
Comment (smathangi) [reply cpo-docker/compose/cpo.yml:23]: i have used one in ccd-docker. But yes, it's a good point.

File: docker-compose.yml [hunk 3]
```diff
@@ -1,26 +1,32 @@
...
-      - $SERVER_PORT:$SERVER_PORT
+      - 4457:4457
+    depends_on:
+      - cpo-db
+
+  cpo-db:
+    image: postgres:11.11
+    restart: always
+    environment:
+      - POSTGRES_DB=cpo
```
Comment (mattnayler) [inline docker-compose.yml:23]: Should this correspond to the default in the `src/main/resources/application.yaml`, i.e. `cpo_case_payment_orders`?

File: docker-compose.yml [hunk 4]
```diff
@@ -1,26 +1,32 @@
...
-      - ROOT_LOGGING_LEVEL
-      - REFORM_SERVICE_TYPE
-      - REFORM_SERVICE_NAME
-      - REFORM_TEAM
-      - REFORM_ENVIRONMENT
-      - LOGBACK_DATE_FORMAT
-      - LOGBACK_REQUIRE_THREAD
-      - LOGBACK_REQUIRE_ALERT_LEVEL=false
-      - LOGBACK_REQUIRE_ERROR_CODE=false
+      SPRING_DATASOURCE_URL: jdbc:postgresql://cpo-db:5432/cpo
```
Comment (mattnayler) [inline docker-compose.yml:9]: Should this correspond to the multiple variables used in the `src/main/resources/application.yaml`?

File: cpo-docker/compose/cpo.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    depends_on:
+      - cpo-db
+    networks:
+      - ccd-network
+
+  cpo-db:
+    image: postgres:11.11
+    restart: always
+    environment:
+      - POSTGRES_DB=cpo
```
Comment (mattnayler) [inline cpo-docker/compose/cpo.yml:26]: see comments in `docker-compose.yml`.

File: cpo-docker/compose/cpo.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+services:
+  case-payment-orders-api:
+    image: "hmctspublic.azurecr.io/cpo/case-payment-orders-api:${CASE_PAYMENT_ORDERS_API_TAG:-latest}"
+    container_name: case-payment-orders-api
+    environment:
+      SERVER_PORT: 4457
+      IDAM_API_URL: "${IDAM_STUB_SERVICE_NAME:-http://idam-api:5000}"
+      IDAM_OIDC_URL: "${IDAM_STUB_SERVICE_NAME:-http://idam-api:5000}"
+      SPRING_DATASOURCE_URL: jdbc:postgresql://cpo-db:5432/cpo
```
Comment (mattnayler) [inline cpo-docker/compose/cpo.yml:12]: see comments in `docker-compose.yml`.

File: cpo-docker/README.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
+# CPO fullstack docker :whale:
```
Comment (mattnayler) [inline cpo-docker/README.md:1]: Is it worth updating/simplifying the _Note_ in the _Functional tests_ section of the main readme by pointing it to this more detailed document on how to run a CPO environment?

**Review by mattnayler — COMMENTED** [review]
Looking good but have added a few questions for clarification.

File: docker-compose.yml (hunk 3 above)
Comment (smathangi) [reply docker-compose.yml:23]: I just used all defaults as it is from application.yaml. We can externalise with params and define few more variables under environment section. But the frequency of changing these default values are very rare.

File: docker-compose.yml (hunk 4 above)
Comment (smathangi) [reply docker-compose.yml:9]: yes, otherwise we need to define all 3 / 4 variables.

File: cpo-docker/README.md (hunk 7 above)
Comment (smathangi) [reply cpo-docker/README.md:1]: that's a good point. will update.

File: docker-compose.yml (hunk 3 above)
Comment (smathangi) [reply docker-compose.yml:23]: Happy to change if parameterisation is required 😄

File: README.md [hunk 8]
```diff
@@ -114,21 +112,14 @@ export TEST_URL=http://localhost:4457
 ./gradlew functional
 ```
 
-> Note: These are the tests run against an environment.  This environment must include access to
-> IDAM and Service Auth Provider services.  All can be made available by installing and running the
-> CCD stack as advised in [ccd-docker](https://github.com/hmcts/ccd-docker), with the minimum
-> enabled services `sidam sidam-local sidam-local-ccd`.
+> Note: These are the tests run against an environment.
+> Please see cpo-docker/README.md for local environment testing.
```
Comment (mattnayler) [inline README.md:116]: Can you use a MD link when linking to other readme file?  
```[cpo-docker/README.md](./cpo-docker/README.md)```

File: docker-compose.yml (hunk 3 above)
Comment (mattnayler) [reply docker-compose.yml:23]: Your value for `POSTGRES_DB` does not match [the equivalent in `application.yaml`](https://github.com/hmcts/cpo-case-payment-orders-api/blob/master/src/main/resources/application.yaml#L36) i.e. `cpo` =! `cpo_case_payment_orders`.

File: docker-compose.yml [hunk 9]
```diff
@@ -1,26 +1,32 @@
...
+
+  cpo-db:
+    image: postgres:11.11
+    restart: always
+    environment:
+      - POSTGRES_DB=cpo
+      - POSTGRES_USER=cpo
+      - POSTGRES_PASSWORD=cpo
+    ports:
+      - 5678:5432
```
Comment (mattnayler) [inline docker-compose.yml:27]: Your external port does not match [the equivalent in `application.yaml`](https://github.com/hmcts/cpo-case-payment-orders-api/blob/master/src/main/resources/application.yaml#L36) i.e. `5678` =! `6968`.

File: docker-compose.yml (hunk 4 above)
Comment (mattnayler) [reply docker-compose.yml:9]: NB: I am keen to but haven't tested your scripts (sorry, I currently cannot as I am mid-edit of my own work).

Is `SPRING_DATASOURCE_URL` a variable that would override [the equivalent in `application.yaml`](https://github.com/hmcts/cpo-case-payment-orders-api/blob/master/src/main/resources/application.yaml#L36)?  I had expected them to be specified separately as per [ccd-def-store](https://github.com/hmcts/ccd-definition-store-api/blob/master/docker-compose.yml#L10-L13).  

I am hoping our end solution allows FTAs to run against your `cpo-docker` but also debug version of CPO running at console or in IDE, I guess still using the database in your docker scripts.

Comment (smathangi) [reply docker-compose.yml:9]: You can override any spring boot variable at any level. SPRING_DATASOURCE_URL is defined by concatenating multiple variables in our app. Either you can override individual ones or override parent variable itself (i.e  SPRING_DATASOURCE_URL). 
The app is up and able to run flyway scripts (that proves DB access).

File: docker-compose.yml (hunk 3 above)
Comment (smathangi) [reply docker-compose.yml:23]: not required it should match with the value in SPRING_DATASOURCE_URL.

File: docker-compose.yml (hunk 9 above)
Comment (smathangi) [reply docker-compose.yml:27]: need not be. all these irrelevant when we re-define " SPRING_DATASOURCE_URL"

Comment (mattnayler) [reply docker-compose.yml:27]: Agreed.  Thanks for the explanation.

File: docker-compose.yml (hunk 4 above)
Comment (mattnayler) [reply docker-compose.yml]: Thanks for the lesson on spring boot variables.  I am happy that your elements will all run together happily.

**Review by mattnayler — APPROVED** [review]
👍

**Review by sullivangareth — APPROVED** [review]

**Pull request merged**